packages/xo-server-auth-ldap/src/index.js

```diff
@@ -250,6 +340,161 @@ class AuthLdap {
+            if (xoGroups.find(group => group.name === name) !== undefined) {
+              logger(`A group called ${name} already exists`)
+              return
+            }
+            xoGroup = await this._xo.createGroup({
+              name,
+              provider: 'ldap',
+              providerGroupId: ldapId,
+            })
+          } else {
+            // Remove it from xoGroups as we will then delete all the remaining
+            // LDAP-provided groups
+            ;[xoGroup] = xoGroups.splice(xoGroupIndex, 1)
+            await this._xo.updateGroup(xoGroup.id, { name })
+          }
+          xoGroup = await this._xo.getGroup(xoGroup.id)
+
+          let ldapGroupMembers = ldapGroup[membersAttribute]
+          ldapGroupMembers = Array.isArray(ldapGroupMembers)
+            ? ldapGroupMembers
+            : [ldapGroupMembers]
+
+          const xoGroupMembers = xoGroup.users.slice(0)
+
+          const actions = []
+          ldapGroupMembers.forEach(ldapId => {
+            const xoUser = xoUsers.find(
+              user => user.authProviders.ldap.id === ldapId
+            )
+            if (xoUser === undefined) {
+              return
+            }
+            // If the user exists in XO, should be a member of the LDAP-provided
+            // group but isn't: add it
+            const userIdIndex = xoGroupMembers.findIndex(id => id === xoUser.id)
+            if (userIdIndex !== -1) {
+              xoGroupMembers.splice(userIdIndex, 1)
+              return
+            }
+
+            actions.push(() => this._xo.addUserToGroup(xoUser.id, xoGroup.id))
+          })
+
+          // All the remaining users of that group can be removed from it since
+          // they're not in the LDAP group
+          actions.push(
+            ...xoGroupMembers.map(userId => () =>
+              this._xo.removeUserFromGroup(userId, xoGroup.id)
+            )
+          )
+          // One by one to avoid race conditions
+          for (const action of actions) {
+            await action()
+          }
+        })
+      )
+
+      // All the remaining groups provided by LDAP can be removed from XO since
+      // they don't exist in the LDAP directory any more
+      await Promise.all(
+        xoGroups
+          .filter(group => group.provider === 'ldap')
+          .map(group => this._xo.deleteGroup(group.id))
+      )
+
+      return groups
+    } finally {
+      await client.unbind()
+    }
+  }
+
+  async _synchronizeUser(ldapUsername, ldapUser) {
+    const { merge, idAttribute } = this._usersConfig
+
+    const ldapId = ldapUser[idAttribute]
+    const xoUsers = await this._xo.getAllUsers()
+    let xoUser = xoUsers.find(
+      xoUser => xoUser.authProviders?.ldap?.id === ldapId
+    )
+    const conflictingXoUser = xoUsers.find(
+      user => user.email === ldapUsername && user.id !== xoUser?.id
+    )
+
+    if (conflictingXoUser !== undefined) {
+      if (xoUser === undefined || merge) {
```

Comment:
In case of an existent `xoUser` and an existent `conflictingXOUser` with a `merge` set to `true`, will have a user which refers to the same `ldapUser`. I think in this case we could merge them.